Added relations between our models

PukieDiederik opened the pull request:
Added the relations between the already existing models. These relations are based on the database diagram in the wiki.

I added the following tables:
- ProjectUser
- CursusUser
- ProjectCursus
- CursusSkill
- ProjectUserCursus
- CursusUserSkill

This should hopefully resolve the following issue:
- #6

pulgamecanica commented:
### Looking Sharp! 💎 

This is looking very nice, but it's missing some very important stuff. 🤔 

***

It's not enough to create the Relationships with the ForeignKeys. We need to specify for each Table, that it has a ManyToMany relation. ⛓️ 
For example, we have to let the Cursus know it has a ManyToMany relation with a User through the CursusUser.
This way, we will be able to fetch all Users from a given Cursus.
```py
c1 = CursusList[1]
c1.give_me_all_the_users()
```

> Of course I am joking with the method name above haha, but we must utilize the Django built-in features to take advantage of this.

It is true we could make queries for this, but we don't want to overwhelm the project with unnecessary queries, and of course it might break compatibility with Django features.

This applies to every other ManyToMany Relation. 👏 

***

Actually Django sometimes omits the intermediate Table for the ManyToMany relation.
If you look at the example on the Django Documentation:

```py
from django.db import models


class Publication(models.Model):
    title = models.CharField(max_length=30)

    class Meta:
        ordering = ["title"]

    def __str__(self):
        return self.title


class Article(models.Model):
    headline = models.CharField(max_length=100)
    publications = models.ManyToManyField(Publication)

    class Meta:
        ordering = ["headline"]

    def __str__(self):
        return self.headline
```

In Theory there should be an Intermediate Table which contains the Relation for the Article and Publisher, something like:
PublishedArticle which would contain the Foreign Key for each member. 🔑 
In reality this table DOES gets created, and if you explore the Database you will find that table, what Django does is an abstraction over this concept.

This is not possible when you want to have extra fields for the intermediate Table, let's say you wanted to also save the date for the PublishedArticle, it would not be possible to do it this way, you would need to create the table Literally, like you did on this PR 🗡️ 

💡 This is just an interesting Fact, you can choose to implement this for the following tables: `ProjectCursus`, `CursusSkill`, `ProjectUserCursus` but it's not mandatory, it's a possibility, also if you rather have it on the models it's ok, and in the future would be easier to implement extra fields or add methods for those tables.

### Tips and Info

[ManyToMany Explained](https://docs.djangoproject.com/en/4.2/topics/db/examples/many_to_many/)
[ManyToMany with extra fields](https://docs.djangoproject.com/en/dev/topics/db/models/#extra-fields-on-many-to-many-relationships)

This two links refer to the documentation with all the info relevant 👯 
I'll try to create some tests to verify this implementation.
So make sure that `User.return_all_cursus()` works.
Or `CursusUser.give_me_my_skills()`

Again I am joking with the methods name, you will find out how Django calls this methods, there are two options, with the plural of the Model Name or with the object set.

pulgamecanica reviewed (changes requested):
Let's implement the ManyToMany relations to make this even better 💯 
Keep it up 🏁

PukieDiederik commented:
Would it be best to revert the migrations and create a new migration when I've implemented the changes. Or is it better to just have a new migration on top of the ones I already have?

pulgamecanica commented:
It doesn't matter, if you prefer a clean migration you can just rollback the migration and delete the file, then do another migration:

```bash
python3 manage.py migrate 0005_skill
rm 0006_cursusskill_cursususer_projectuser_projectusercursus_and_more
rm 0007_projectuser_intra_id
```
Then you can make changes on the models and migrate again ;)

PukieDiederik commented:
I added many to many relations to our models using the `ManyToManyField` from django.

I also changed foreign key to use strings instead of the models to prevent NameErrors. We might get name errors because some classes aren't defined yet when we would be using them. And for some it's impossible to move them around because they have a circular dependency.

pulgamecanica commented on backend/portfolio42_api/models.py:
```diff
@@ -49,6 +52,9 @@ class Project(models.Model):
 	updated_at = models.DateTimeField(auto_created=True, auto_now=True)
+
+	users = models.ManyToManyField('User', through='ProjectUser')
+	projects = models.ManyToManyField('Project', through='ProjectCursus')
```
This should be cursus, & Cursus, instead of projects & Project

pulgamecanica commented on backend/portfolio42_api/models.py:
```diff
@@ -66,4 +72,50 @@ def __str__(self):
+	finished_at = models.DateTimeField()
+	updated_at = models.DateTimeField(auto_created=True, auto_now=True)
+
+	cursus = models.ManyToManyField('Cursus', through='ProjectUserCursus')
```
We are going to remove this, the cursus should only be dependant on the Project, not the ProjectUser.

pulgamecanica commented on backend/portfolio42_api/models.py:
```diff
@@ -66,4 +72,50 @@ def __str__(self):
+	id_skill = models.ForeignKey('Skill', on_delete=models.CASCADE)
+
+# Creates a relation between ProjectUser and Cursus, it relates which project users are part of which cursus
+class ProjectUserCursus(models.Model):
```
We should remove this class, not needed anymore ... my bad :crying_cat_face:

pulgamecanica reviewed (changes requested):
# Changes :cactus: 
Hey, let's discuss removing the ProjectUserCursus class.
***
Also change the projects to cursus on the Project class.

### Cheers :wine_glass:

pulgamecanica commented:
## DB proposal (visual)
***

<img width="1477" alt="Screenshot 2023-05-18 at 12 15 12" src="https://github.com/pulgamecanica/42Portfolio/assets/28810331/8bdd7d6b-4394-4357-9df7-8458c386fb89">

pulgamecanica commented on backend/portfolio42_api/models.py:
```diff
@@ -35,6 +35,9 @@ class Cursus(models.Model):
 
+	skills = models.ManyToManyField('Skill', through='CursusSkill')
+	users = models.ManyToManyField('User', through='CursusUser')
+
```
Here we should add projects which has a relationship Many to Many with the cursus.

PukieDiederik replied:
So basically add manytomany on both parts of the relation?

pulgamecanica replied:
Yes, when need access from both sides.
cursusObj1.projects => returns list of projects
projectObj1.cursus => returns list of cursus

pulgamecanica reviewed (changes requested):
## User Relations

We should add the User relations as well.

 The Users have: 

> Many Cursus. (Through CursusUser)
>
> Many Projects. (Through ProjectUser)
>

***

## Tips

The skills don't belong directly to the user, they belong to a CursusUser that is why they are not included in this list, which the logic for this has been done already.

***

pulgamecanica reviewed (approved):
Awesome work!
🥂 Keep it up.
